core/src/main/scala/kafka/server/DelayedProduce.scala

```diff
@@ -58,6 +62,8 @@ class DelayedProduce(delayMs: Long,
                      lockOpt: Option[Lock] = None)
   extends DelayedOperation(delayMs, lockOpt) {
 
+  override lazy val logger = DelayedProduce.logger
```

Comment:
Instead of doing this, you can simply import the logger from the companion object.

Reply:
Oh, if the superclass needs this, then we can keep as is but change it to a def.

Reply:
Thanks for reviewing this!

If I understand this correctly, I believe you're suggesting to replace the declaration in the [Logging trait](https://github.com/apache/kafka/blob/5c2492bca71200806ccf776ea31639a90290d43e/core/src/main/scala/kafka/utils/Logging.scala#L43) to be a `protected def logger = Logger(LoggerFactory.getLogger(loggerName))` instead of `protected lazy val logger = Logger(LoggerFactory.getLogger(loggerName))`

I think that will result in a logger being instantiated for each usages of `logger` which are *not* overridden by a `lazy val`. This is something I reckon we should avoid to exacerbate the problem. [This example](https://scastie.scala-lang.org/X2OWyu5BRJKXRA81zxsgJw) in the playground demonstrates the behaviour.

Please let me know your thoughts and if there's something I overlooked or misunderstood :)

Reply:
No, I'm saying to change this lazy val to be a def. The logger it will reference is a val in the companion object.

Reply:
That fails compilation with the following error:

```
stable, immutable value required to override:
lazy protected val logger: com.typesafe.scalalogging.Logger (defined in trait Logging)
```

I suppose we cannot override a `lazy val` with a `def`.

Example: https://scastie.scala-lang.org/qI1plzLAQsC10hYbEBHk5A

Reply:
@ijuma any thoughts on the above? Perhaps I'm overlooking something and it would help if you could elaborate with an example on scastie.

Reply:
@ijuma Wonder if you got a chance to revisit this? I've also found some [precedent](https://github.com/apache/kafka/blob/959f9ca4c02817cce85f0375d431ed72f22699e2/core/src/main/scala/kafka/controller/StateChangeLogger.scala#L37) where a `lazy val` is used for overriding the logger.

Reply:
Yes, you are right - I had overlooked that the Logging trait defined `logger` as lazy val already since it provides a default implementation.